File: dui/src/utils/proxy.ts
import { invoke } from '@tauri-apps/api/core';

export interface ProxyInfo {
    port: number;
    target: string;
    is_running: boolean;
}

export async function getProxyInfo(): Promise<ProxyInfo> {
    // Add type assertion to ensure we get the expected shape
    type ProxyResponse = { port: number; target: string; };
    console.debug('getProxyInfo function called');
    console.debug('Preparing to invoke get_proxy_info command...');
    try {
        console.debug('Sending invoke call to get_proxy_info...');
        const response = await invoke<ProxyResponse>('get_proxy_info');
        console.debug('Invoke call completed');
        console.debug('Raw response from backend:', {
            received: response,
            type: typeof response,
            keys: response ? Object.keys(response) : 'null'
        });
        
        // Validate response shape
        if (!response || typeof response.port !== 'number' || typeof response.target !== 'string') {
            console.error('Invalid proxy info response:', {
                response,
                portType: response ? typeof response.port : 'undefined',
                targetType: response ? typeof response.target : 'undefined'
            });
            throw new Error('Invalid proxy info response format');
        }
        
        const info: ProxyInfo = {
            port: response.port,
            target: response.target
        };
        console.debug('Received proxy info from backend:', info);
        return info;
    } catch (error) {
        console.error('Failed to get proxy info:', error);
        console.debug('Error details:', {
            message: error.message,
            stack: error.stack,
            error
        });
        throw error;
    }
}

export async function setDebugMode(debug_mode: boolean): Promise<void> {
    console.debug('Setting debug mode to:', debug_mode);
    return invoke('set_debug_mode', { debug_mode });
}

export async function setProxyTarget(target: string): Promise<void> {
    console.debug('Setting proxy target to:', target);
    return invoke('set_proxy_target', { target });
}

// Helper function to determine if proxy should be used
export function shouldUseProxy(useTls: boolean): boolean {
    const useProxy = !useTls;
    console.debug(`Should use proxy? ${useProxy} (TLS: ${useTls})`);

    return !useTls; // Only use proxy when TLS is disabled
}

// Start proxy server function to match API/BUI pattern
export async function startProxyServer(): Promise<{ success: boolean; error?: string }> {
    console.debug('Starting proxy server...');
    try {
        await invoke('start_proxy_server');
        console.debug('Proxy server started successfully');
        return { success: true };
    } catch (error) {
        console.error('Failed to start proxy server:', error);
        return { 
            success: false, 
            error: (error instanceof Error) ? (error.message || 'Unknown error') : error 
        };
    }
}

// Stop proxy server function to match API/BUI pattern
export async function stopProxyServer(): Promise<{ success: boolean; error?: string }> {
    console.debug('Stopping proxy server...');
    try {
        await invoke('stop_proxy_server');
        console.debug('Proxy server stopped successfully');
        return { success: true };
    } catch (error) {
        console.error('Failed to stop proxy server:', error);
        return { 
            success: false, 
            error: (error instanceof Error) ? (error.message || 'Unknown error') : error 
        };
    }
}

File: dui/src-tauri/src/commands/proxy.rs
use crate::proxy::HttpProxy;
use log::{debug, info};
use std::sync::Arc;
use tokio::sync::RwLock;

#[tauri::command]
pub async fn get_proxy_info(
    state: tauri::State<'_, Arc<RwLock<HttpProxy>>>,
) -> Result<crate::proxy::ProxyInfo, String> {
    debug!("get_proxy_info command invoked");
    let proxy = state.read().await;
    let target = proxy.target_url.read().await.clone();
    let is_running = proxy.is_running().await;

    Ok(crate::proxy::ProxyInfo {
        port: proxy.port,
        target,
        is_running,
    })
}

#[tauri::command]
pub async fn start_proxy_server(
    state: tauri::State<'_, Arc<RwLock<HttpProxy>>>,
) -> Result<(), String> {
    debug!("start_proxy_server command invoked");
    let proxy = state.read().await;
    proxy.start().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stop_proxy_server(
    state: tauri::State<'_, Arc<RwLock<HttpProxy>>>,
) -> Result<(), String> {
    debug!("stop_proxy_server command invoked");
    let proxy = state.read().await;
    proxy.stop().await
}

#[tauri::command]
pub async fn set_debug_mode(
    debug_mode: bool,
    state: tauri::State<'_, Arc<RwLock<HttpProxy>>>,
) -> Result<(), String> {
    debug!("set_debug_mode called with debug_mode: {}", debug_mode);
    let proxy = state.read().await;
    *proxy.debug_mode.write().await = debug_mode;
    debug!("Successfully updated debug mode to: {}", debug_mode);
    Ok(())
}

#[tauri::command]
pub async fn set_proxy_target(
    target: String,
    state: tauri::State<'_, Arc<RwLock<HttpProxy>>>,
) -> Result<(), String> {
    debug!("set_proxy_target called with target: {}", target);
    // Validate target URL
    let parsed_url =
        reqwest::Url::parse(&target).map_err(|e| format!("Invalid target URL: {}", e))?;

    if parsed_url.scheme() != "https" {
        return Err(format!(
            "Invalid URL scheme: {}. Only HTTPS URLs are allowed.",
            parsed_url.scheme()
        ));
    }

    debug!(
        "Parsed target URL - scheme: {}, host: {:?}, port: {:?}",
        parsed_url.scheme(),
        parsed_url.host_str(),
        parsed_url.port()
    );

    debug!("Setting proxy target to: {}", target);
    let proxy = state.read().await;
    *proxy.target_url.write().await = target.clone();
    debug!("Successfully updated proxy target to: {}", target);
    info!("Proxy target updated to: {}", target);
    Ok(())
}
